File: lib/Core/services/notification_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_messaging/firebase_messaging.dart';
import 'package:flutter_local_notifications/flutter_local_notifications.dart';

class NotificationService {
  static final NotificationService _instance = NotificationService._internal();
  factory NotificationService() => _instance;
  NotificationService._internal();

  final FirebaseFirestore _firestore = FirebaseFirestore.instance;
  final FlutterLocalNotificationsPlugin _localNotifications = FlutterLocalNotificationsPlugin();

  // Khởi tạo service
  Future<void> initialize() async {
    // Cấu hình local notifications
    const AndroidInitializationSettings initializationSettingsAndroid =
        AndroidInitializationSettings('@mipmap/ic_launcher');
    
    const DarwinInitializationSettings initializationSettingsIOS =
        DarwinInitializationSettings();
    
    const InitializationSettings initializationSettings = InitializationSettings(
      android: initializationSettingsAndroid,
      iOS: initializationSettingsIOS,
    );

    await _localNotifications.initialize(initializationSettings);

    // Cấu hình Android channel
    const AndroidNotificationChannel channel = AndroidNotificationChannel(
      'high_importance_channel',
      'High Importance Notifications',
      description: 'This channel is used for important notifications.',
      importance: Importance.high,
    );

    await _localNotifications
        .resolvePlatformSpecificImplementation<AndroidFlutterLocalNotificationsPlugin>()
        ?.createNotificationChannel(channel);

    // Xử lý thông báo khi app đang mở
    FirebaseMessaging.onMessage.listen((RemoteMessage message) {
      _handleForegroundMessage(message);
    });

    // Xử lý khi user tap vào thông báo
    FirebaseMessaging.onMessageOpenedApp.listen((RemoteMessage message) {
      _handleNotificationTap(message);
    });
  }

  // Lưu thông báo vào Firebase Database
  Future<void> saveNotificationToDatabase({
    required String title,
    required String body,
    required String userId,
    required String type,
    String? orderId,
    String? status,
  }) async {
    try {
      final notificationData = {
        'title': title,
        'body': body,
        'userId': userId,
        'type': type,
        'orderId': orderId,
        'status': status,
        'read': false,
        'createdAt': FieldValue.serverTimestamp(),
      };

      final DocumentReference docRef = await _firestore.collection('notifications').add(notificationData);
      print('Notification saved to Firebase Database: ${docRef.id}');
    } catch (e) {
      print('Error saving notification to Firebase Database: $e');
    }
  }

  // Xử lý thông báo khi app đang mở
  void _handleForegroundMessage(RemoteMessage message) async {
    print('Got a message whilst in the foreground!');
    print('Message data: ${message.data}');

    if (message.notification != null) {
      print('Message also contained a notification: ${message.notification}');
      
      // Hiển thị local notification
      await _showLocalNotification(message);
      
      // Lưu vào Firebase Database
      await saveNotificationToDatabase(
        title: message.notification?.title ?? 'Thông báo',
        body: message.notification?.body ?? '',
        userId: message.data['userId'] ?? '',
        type: message.data['type'] ?? 'general',
        orderId: message.data['orderId'],
        status: message.data['status'],
      );
    }
  }

  // Xử lý khi user tap vào thông báo
  void _handleNotificationTap(RemoteMessage message) {
    print('Notification tapped: ${message.data}');
    // Có thể thêm logic điều hướng dựa trên loại thông báo
    if (message.data['type'] == 'order_created') {
      // Điều hướng đến trang đơn hàng
    } else if (message.data['type'] == 'order_confirmed') {
      // Điều hướng đến trang đơn hàng
    }
  }

  // Hiển thị local notification
  Future<void> _showLocalNotification(RemoteMessage message) async {
    const AndroidNotificationDetails androidPlatformChannelSpecifics =
        AndroidNotificationDetails(
      'high_importance_channel',
      'High Importance Notifications',
      channelDescription: 'This channel is used for important notifications.',
      importance: Importance.max,
      priority: Priority.high,
      showWhen: false,
    );
    
    const NotificationDetails platformChannelSpecifics =
        NotificationDetails(android: androidPlatformChannelSpecifics);
    
    await _localNotifications.show(
      message.hashCode,
      message.notification?.title,
      message.notification?.body,
      platformChannelSpecifics,
      payload: message.data.toString(),
    );
  }

  // Lấy danh sách thông báo của user
  Future<List<Map<String, dynamic>>> getUserNotifications(String userId) async {
    try {
      final QuerySnapshot snapshot = await _firestore
          .collection('notifications')
          .where('userId', isEqualTo: userId)
          .orderBy('createdAt', descending: true)
          .get();

      List<Map<String, dynamic>> notifications = [];
      
      for (var doc in snapshot.docs) {
        final data = doc.data();
        if (data != null) {
          notifications.add({
            'id': doc.id,
            ...Map<String, dynamic>.from(data as Map<String, dynamic>),
          });
        }
      }

      return notifications;
    } catch (e) {
      print('Error getting user notifications: $e');
      return [];
    }
  }

  // Đánh dấu thông báo đã đọc
  Future<void> markNotificationAsRead(String notificationId) async {
    try {
      await _firestore.collection('notifications').doc(notificationId).update({'read': true});
    } catch (e) {
      print('Error marking notification as read: $e');
    }
  }

  // Xóa thông báo
  Future<void> deleteNotification(String notificationId) async {
    try {
      await _firestore.collection('notifications').doc(notificationId).delete();
    } catch (e) {
      print('Error deleting notification: $e');
    }
  }

  // Lấy số thông báo chưa đọc
  Future<int> getUnreadNotificationCount(String userId) async {
    try {
      final QuerySnapshot snapshot = await _firestore
          .collection('notifications')
          .where('userId', isEqualTo: userId)
          .where('read', isEqualTo: false)
          .get();

      return snapshot.docs.length;
    } catch (e) {
      print('Error getting unread notification count: $e');
      return 0;
    }
  }
} 

File: flash_food_backend/services/notificationService.js
const admin = require('firebase-admin');
const serviceAccount = require('../firebase-service-account.json');

// Khởi tạo Firebase Admin SDK nếu chưa có
if (!admin.apps.length) {
  admin.initializeApp({
    credential: admin.credential.cert(serviceAccount)
  });
}

const db = admin.firestore();

class NotificationService {
  // Lưu thông báo vào Firebase Database
  static async saveNotification({
    title,
    body,
    userId,
    type,
    orderId,
    status,
    reason = null
  }) {
    try {
      const notificationData = {
        title: title,
        body: body,
        userId: userId,
        type: type,
        orderId: orderId,
        status: status,
        read: false,
        createdAt: admin.firestore.FieldValue.serverTimestamp(),
        ...(reason && { reason: reason })
      };

      const docRef = await db.collection('notifications').add(notificationData);
      console.log('Notification saved to Firebase Database with ID:', docRef.id);
      return docRef.id;
    } catch (error) {
      console.error('Error saving notification to Firebase Database:', error);
      throw error;
    }
  }

  // Tạo thông báo đặt hàng thành công
  static async createOrderSuccessNotification({
    userId,
    orderId,
    orderTotal,
    orderItems = []
  }) {
    try {
      const title = 'Đặt hàng thành công! 🎉';
      const body = `Đơn hàng #${orderId} của bạn đã được đặt thành công với tổng tiền ${orderTotal} VND. Chúng tôi sẽ xử lý đơn hàng của bạn sớm nhất!`;
      
      const notificationData = {
        title: title,
        body: body,
        userId: userId,
        type: 'order_success',
        orderId: orderId,
        status: 'pending',
        orderTotal: orderTotal,
        orderItems: orderItems,
        read: false,
        createdAt: admin.firestore.FieldValue.serverTimestamp(),
      };

      const docRef = await db.collection('notifications').add(notificationData);
      console.log('Order success notification saved with ID:', docRef.id);
      return docRef.id;
    } catch (error) {
      console.error('Error creating order success notification:', error);
      throw error;
    }
  }

  // Tạo thông báo trạng thái đơn hàng
  static async createOrderStatusNotification({
    userId,
    orderId,
    status,
    message
  }) {
    try {
      let title = '';
      let body = '';
      
      switch (status) {
        case 'confirmed':
          title = 'Đơn hàng đã được xác nhận! ✅';
          body = `Đơn hàng #${orderId} đã được xác nhận và đang được chuẩn bị.`;
          break;
        case 'preparing':
          title = 'Đơn hàng đang được chuẩn bị! 👨‍🍳';
          body = `Đơn hàng #${orderId} đang được chuẩn bị. Sẽ sẵn sàng sớm!`;
          break;
        case 'ready':
          title = 'Đơn hàng đã sẵn sàng! 🚚';
          body = `Đơn hàng #${orderId} đã sẵn sàng và đang được giao đến bạn.`;
          break;
        case 'delivered':
          title = 'Đơn hàng đã được giao! 🎉';
          body = `Đơn hàng #${orderId} đã được giao thành công. Cảm ơn bạn đã sử dụng dịch vụ của chúng tôi!`;
          break;
        case 'cancelled':
          title = 'Đơn hàng đã bị hủy! ❌';
          body = `Đơn hàng #${orderId} đã bị hủy. ${message || 'Vui lòng liên hệ với chúng tôi nếu có thắc mắc.'}`;
          break;
        default:
          title = 'Cập nhật đơn hàng';
          body = message || `Đơn hàng #${orderId} có cập nhật mới.`;
      }

      const notificationData = {
        title: title,
        body: body,
        userId: userId,
        type: 'order_status',
        orderId: orderId,
        status: status,
        message: message,
        read: false,
        createdAt: admin.firestore.FieldValue.serverTimestamp(),
      };

      const docRef = await db.collection('notifications').add(notificationData);
      console.log('Order status notification saved with ID:', docRef.id);
      return docRef.id;
    } catch (error) {
      console.error('Error creating order status notification:', error);
      throw error;
    }
  }

  // Lấy danh sách thông báo của user
  static async getUserNotifications(userId) {
    try {
      // Chỉ filter theo userId, không orderBy để tránh cần index phức tạp
      const snapshot = await db.collection('notifications')
        .where('userId', '==', userId)
        .get();

      const notifications = [];
      snapshot.forEach((doc) => {
        notifications.push({
          id: doc.id,
          ...doc.data()
        });
      });

      // Sort trong JavaScript thay vì trong query
      notifications.sort((a, b) => {
        const dateA = a.createdAt ? new Date(a.createdAt.toDate()) : new Date(0);
        const dateB = b.createdAt ? new Date(b.createdAt.toDate()) : new Date(0);
        return dateB - dateA; // Descending order
      });

      return notifications;
    } catch (error) {
      console.error('Error getting user notifications:', error);
      return [];
    }
  }

  // Đánh dấu thông báo đã đọc
  static async markNotificationAsRead(notificationId) {
    try {
      await db.collection('notifications').doc(notificationId).update({
        read: true
      });
      console.log('Notification marked as read:', notificationId);
    } catch (error) {
      console.error('Error marking notification as read:', error);
      throw error;
    }
  }

  // Xóa thông báo
  static async deleteNotification(notificationId) {
    try {
      await db.collection('notifications').doc(notificationId).delete();
      console.log('Notification deleted:', notificationId);
    } catch (error) {
      console.error('Error deleting notification:', error);
      throw error;
    }
  }

  // Lấy số thông báo chưa đọc
  static async getUnreadNotificationCount(userId) {
    try {
      const snapshot = await db.collection('notifications')
        .where('userId', '==', userId)
        .where('read', '==', false)
        .get();

      return snapshot.size;
    } catch (error) {
      console.error('Error getting unread notification count:', error);
      return 0;
    }
  }
}

module.exports = NotificationService; 